model/flow/identity.go

```diff
@@ -508,3 +561,99 @@ func (il IdentityList) EqualTo(other IdentityList) bool {
 	}
 	return true
 }
+
+// Exists takes a previously sorted Identity list and searches it for the target value
+// This code is optimized, so the coding style will be different
+// target:  value to search for
+func (il IdentityList) Exists(target *Identity) bool {
+	left := 0
+	lenList := len(il)
+	right := lenList - 1
+	mid := int(uint(right) >> 1)
+	num2 := target.NodeID[:]
+
+	// pre-calculate these 4 values for comparisons later
+	var tgt [4]uint64
+	tgt[0] = binary.BigEndian.Uint64(num2[:])
+	tgt[1] = binary.BigEndian.Uint64(num2[8:])
+	tgt[2] = binary.BigEndian.Uint64(num2[16:])
+	tgt[3] = binary.BigEndian.Uint64(num2[24:])
+
+	for {
+		num1 := il[mid].NodeID[:]
+		lenID := len(num1)
+		i := 0
+
+		// assume the length is a multiple of 8, for performance.  it's 32 bytes
+		for {
+			chunk1 := binary.BigEndian.Uint64(num1[i:])
+			chunk2 := tgt[i/8]
+
+			if chunk1 < chunk2 {
+				left = mid + 1
+				break
+			} else if chunk1 > chunk2 {
+				right = mid - 1
+				break
+			} else if i >= lenID-8 {
+				// we're on the last chunk of 8 bytes, and
+				// so return true if equal -- it exists
+				return true
+			}
+
+			// these 8 bytes were equal, so increment index by 8 bytes
+			i += 8
+		}
+		if left > right {
+			return false
+		}
+		mid = int(uint(left+right) >> 1)
+	}
+}
+
+// Exists takes a previously sorted Identity list and searches it for the target value
+// target:  value to search for
+func (il IdentityList) IdentifierExists(target Identifier) bool {
```

Comment:
In general I'm a bit worried about accidental misuse of these methods which require the receiver identity list to be ordered. It isn't obvious to me from the function name or receiver type that the receiver needs to be sorted. Could we do 1 or 1+2 of these:

1. Make the requirement a bit "louder" in the function documentation, and be explicit about the expected sort order:
```
// documentation ...
//
// CAUTION: IdentityList receiver must be previously sorted lexicographically by node ID in order to use this function.
// <link to function which sorts an identity list to the required order>
func ...
```

2. Introduce a type representing a sorted list that must be used when we require a receiver to be sorted already.
```go
type SortedIdentityList IdentityList

func (il IdentityList) Sort() SortedIdentityList { ... }

func (il SortedIdentityList) IdentifierExists(nodeID Identifier) bool { ... }
// other functions which require a sorted receiver
```

Reply:
will do